### DISCUSSION TITLE
HR Module Design Discussions

### DISCUSSION THREAD

**Discussion by neilt** [ledgersmb/LedgerSMB#8087] [Design]
Discussion for enhancements to the Human Resource (HR) features. Please feel free to comment or add.

Related Issues and Discussions:

1. https://github.com/ledgersmb/LedgerSMB/discussions/7879
2. https://github.com/freelock/payroll-app
3. https://www.ksrevenue.gov/faqs-withholding.html

Summary of Element discussion on 19 March 2024:
1. Additional HR features would be nice to have
2. @freelock has experience with an [ older module](https://github.com/freelock/payroll-app) that needs updating. This might be the best way to get initial wage tax calculations into LSMB.
3. The probable implementation path is to implement a local wage tax calculation module for tax calculations involving a single location directly in LSMB. LSMB should implement the tax calculations using a plugin or other infrastructure that makes it easy to implement locally.  This initial solution may not handle remote work.
5. Once an organization's employees need withholding from multiple locations (cities, states, provinces, etc) LSMB will probably be better off interfacing with a wage and tax processing service through an API.

19 Mar 2024 - Added related item 3 in reference to @freelock comment below.

**Comment by freelock** [comment]
To clarify, my payroll system is not (yet) associated with LedgerSMB. I was hoping to use it to post data directly into LedgerSMB, and did a fair amount of modeling of the problem to get it working for our needs (for Washington State, which does not have an income tax).

Based on that, I can outline some of the needs...

First of all, each deduction needs to have its own model -- a specific set of accounts, and plugin/algorithm to follow.

In my system, I have a list of raw taxes, which are linked to the type:

- Medicaid/Medicare (employer portion, employee portion) - flat rate pre-deductions
- Social Security (employer portion, employee portion each need liability accounts). Flat rate pre-deductions with a cap
- Retirement deductions - employee, employer match
- Health care employee deduction
- Federal Withholding (custom calculations, needs updating every year) - basis is post-deduction
- WA Labor and Industries (Workers Comp) - flat rate on hours worked, employer and employee portions
- Employment Security (unemployment) - flat rate on total hours, employer and employee portions
- WA Family/Medical Leave - flat rate on total hours, employer and employee portions
- WA Cares - flat rate on total hours, employee withholding

Each of these needs a "Basis" in terms of hours or pay, including PTO/Holiday or excluding, including deductions or some deductions or not, and with a maximum amount or not -- all based on each employee's hours and pay.

Each of these needs at least one and often two liability accounts.

For each employee, I have a configuration for whether or not each tax applies, and other options (such as W-2 type, extra deductions, etc)

For each employee, we also have hour categories for Regular, Overtime, Holiday, PTO, and a "bonus" value in dollars instead of hours.

For deductions, need to track PTO balance, retirement contributions, totals over time periods (pay period, month, quarter, year).

There's lots of variation from there for the specific details on amounts -- e.g. for Neil's examples related to non-resident employees in Kansas, but I think those can be handled by this basic architecture, with plugins?

So the next question might be, what does the API look like for a payroll service? And does that map to these basic units?
